File: src/4node/apply/nodes/confirmNode.ts
// ─── confirmNode — 确认路由 + 确认节点 ──────────────────────────────────────
// 归属：apply 子图
// confirmRoute：Router，判断是否有匹配结果
// confirmNode：Node，interrupt 等待前端确认

import { HumanMessage, AIMessage } from '@langchain/core/messages'
import { interrupt } from '@langchain/langgraph'
import type { ApplyStateType } from '../../../3state/state.js'
import { parseCheckResults } from '../utils.js'

// ── Router ───────────────────────────────────────────────────────────────

/**
 * 路由判断：有匹配结果 → 进入 confirm 等待用户确认；无结果 → 直接结束
 */
export function confirmRoute(state: ApplyStateType): 'confirm' | 'end' {
  const suggestions = parseCheckResults(state.checkResults)
  return suggestions.length > 0 ? 'confirm' : 'end'
}

// ── Node ─────────────────────────────────────────────────────────────────

/**
 * 等待用户确认并上传证明材料（interrupt）
 *
 * 通过 interrupt 暂停 LangGraph，等待前端：
 *   - 用户点击「确认提交」→ 传入 { action: 'confirm', proofFileIds: [...] }
 *   - 用户点击「取消」→ 传入 { action: 'cancel' }
 */
export async function confirmNode(
  state: ApplyStateType
): Promise<Partial<ApplyStateType>> {
  console.log('--apply:confirm (interrupt)')

  const suggestions = parseCheckResults(state.checkResults)

  const question = [
    `已为您匹配到以下加分项，请上传对应证明材料后点击「确认提交」：`,
    '',
    ...suggestions.map((s: any, i: number) =>
      `${i + 1}. **${s.templateName}** / ${s.ruleName}（预计 ${s.estimatedScore} 分）\n ${s.reason}`
    ),
  ].join('\n')

  const userAnswer = interrupt({ type: 'confirm' as const, question, suggestions })

  return {
    messages: [
      new AIMessage(question),
      new HumanMessage(String(userAnswer)),
    ],
  }
}

File: src/nodes/apply/confirm_node.py
"""confirmNode - 确认路由 + 确认节点 - 对应 TS: confirmNode.ts"""
import json
from typing import Literal
from langgraph.types import interrupt, Command
from langchain_core.messages import HumanMessage, AIMessage
from .summarize_node import parse_check_results


def confirm_route(state: dict) -> Literal["confirm", "__end__"]:
    """
    路由判断: 有匹配结果 → confirm；无结果 → 结束
    """
    check_results: list[str] = state.get("check_results", [])
    suggestions = parse_check_results(check_results)
    return "confirm" if suggestions else "__end__"


async def confirm_node(state: dict) -> dict:
    """
    确认节点
    
    归属: apply 子图
    功能: 等待用户确认并上传证明材料（使用 interrupt）
    """
    print("--apply:confirm (interrupt)")
    
    check_results: list[str] = state.get("check_results", [])
    suggestions = parse_check_results(check_results)
    
    # 生成确认问题
    question_lines = ["已为您匹配到以下加分项，请上传对应证明材料后点击「确认提交」：", ""]
    for i, s in enumerate(suggestions, 1):
        question_lines.append(f"{i}. **{s['templateName']}** / {s['ruleName']}（预计 {s['estimatedScore']} 分）")
        question_lines.append(f"   {s['reason']}")
        question_lines.append("")
    
    question = "\n".join(question_lines)
    
    # 使用 interrupt 暂停，等待用户确认
    user_answer = interrupt({
        "type": "confirm",
        "question": question,
        "suggestions": suggestions
    })
    
    return {
        "messages": [
            AIMessage(content=question),
            HumanMessage(content=str(user_answer))
        ]
    }
